File: 100-Days-Series/Day 99/PermutationsOfAString.java
public class PermutationsOfAString {
    // Function to swap characters in a string
    private static void swap(char[] str, int i, int j) {
        char temp = str[i];
        str[i] = str[j];
        str[j] = temp;
    }
    // Recursive function to generate permutations
    private static void permute(char[] str, int l, int r) {
        if (l == r) {
            System.out.println(String.valueOf(str)); // Print the permutation
        } else {
            for (int i = l; i <= r; i++) {
                swap(str, l, i); // Swap current character with the leftmost character
                permute(str, l + 1, r); // Recur for the rest of the string
                swap(str, l, i); // Backtrack to the original configuration
            }
        }
    }
    // Main Method
    public static void main(String[] args) {
        // Let's Start Here
        String str = "ABC";
        char[] charArray = str.toCharArray(); // Convert string to character array
        System.out.println("All permutations of the string \"" + str + "\" are:");
        permute(charArray, 0, charArray.length - 1);
    }
}

File: 100-Days-Series/Day 99/PermutationsOfAString.cpp
#include <iostream>
#include <string>
using namespace std;
// Function to swap two characters
void swap(char &x, char &y) {
    char temp = x;
    x = y;
    y = temp;
}
// Recursive function to generate permutations
void permute(string &str, int l, int r) {
    if (l == r) {
        cout << str << endl; // Print the permutation
    } else {
        for (int i = l; i <= r; i++) {
            swap(str[l], str[i]); // Swap current character with the leftmost character
            permute(str, l + 1, r); // Recur for the rest of the string
            swap(str[l], str[i]); // Backtrack to the original configuration
        }
    }
}
// Let's Start Here
int main() {
    string str = "ABC";
    cout << "All permutations of the string \"" << str << "\" are:" << endl;
    permute(str, 0, str.length() - 1);
    return 0;
}
